File: 2018/Day 6/Solution.py
import re, collections
with open('input.txt') as f:
    parse = lambda x: tuple(map(int, re.findall(r'\d+', x)))
    points = [*map(parse, f.read().splitlines())]
result = collections.defaultdict(int)
a = min(x for x,_ in points)
b = min(y for _,y in points)
c = max(x for x,_ in points)
d = max(y for _,y in points)
corners = [(a,b),(c,b),(a,d),(c,d)]
fuck2 = int()
def m_dist(x1, y1, x2, y2):
    return abs(x1-x2) + abs(y1-y2)
def maybe(x, y):
    ll = [(m_dist(x,y,*p), p) for p in points+corners]
    val, p = min(ll, key = lambda x: x[0])
    if [v[0] for v in ll].count(val) == 1 and p not in corners:
        result[p] += 1
def eat_shit(x, y):
    global fuck2
    ll = sum([m_dist(x,y,*p) for p in points])
    if ll < 10000:
        fuck2 += 1
for y in range(b,d):
    for x in range(a,c):
        maybe(x, y)
        eat_shit(x, y)
print(max(result.values()), fuck2)

File: 2018/Day 6/Solution.hs
{-#LANGUAGE PartialTypeSignatures#-}
module Main where

import qualified Data.Map.Lazy as Map
import Control.Monad.State
import Data.List.Split
import Data.Function
import Text.Printf
import Data.List

type Counter = Map.Map Coords Int
type Coords = (Int, Int)

unpair (a,b) = [a,b]
pair [a,b] = (a,b)

main :: IO()
main = do
  points <- input
  let start = (points, Map.empty, 0)
  let ranges = pair <$> (mapM range.unpair.unzip) points
  let result = distance ranges `evalState` start
  printf "part 1: %d\n" $fst result
  printf "part 2: %d\n" $snd result

input :: IO [Coords]
input = map (pair.parse).lines<$>readFile "input.txt"
  where parse = map read.splitOn ","

distance :: [Coords] -> State ([Coords], Counter, Int) (Int, Int)
distance [] = gets yield where
  yield (_, m, t) = (maximum $Map.elems m, t)
distance (p:next) = do
  (points, tracking, total) <- get
  put (points, case nearest points p of
    Just match -> Map.insertWith (+) match 1 tracking
    Nothing -> tracking, if closeby points p then
      total + 1 else total); distance next

nearest :: [Coords] -> Coords -> Maybe Coords
nearest points p = do
  let table = m_dist' p <$> points
  let (val, key) = minimumBy (on compare fst) table
  if count (map fst table) val == 1 then Just key else Nothing

closeby :: [Coords] -> Coords -> Bool
closeby points p = 10000 > (sum $m_dist p <$> points)

m_dist :: Coords -> Coords -> Int
m_dist (x1,y1) (x,y) = on (+) abs (x1-x) (y1-y)

m_dist' :: Coords -> Coords -> (Int, Coords)
m_dist' a b = (m_dist a b, b)

range :: (Foldable t, Ord a, Num a, Enum a) => t a -> [a]
range ys = [minimum ys..maximum ys-1]

count :: Eq a => [a] -> a -> Int
count xs x = length $filter (== x) xs
